tests/unit/test_repo.py

```diff
@@ -146,3 +151,292 @@ def test_can_add_repositories_from_string_with_options(self):
             "deb [arch=amd64 signed-by=/foo/gpg.key] https://example.com/foo focal bar baz\n",
             open(d.filename).readlines(),
         )
+
+
+ubuntu_sources_deb822 = """
+Types: deb
+URIs: http://nz.archive.ubuntu.com/ubuntu/
+Suites: noble noble-updates noble-backports
+Components: main restricted universe multiverse
+Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
+
+Types: deb
+URIs: http://security.ubuntu.com/ubuntu
+Suites: noble-security
+Components: main restricted universe multiverse
+Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
+"""
+
+ubuntu_sources_deb822_with_comments = """
+Components: main restricted universe multiverse  # this lib doesn't care about order
+Types: deb  # this could include deb-src as well or instead
+URIs: http://nz.archive.ubuntu.com/ubuntu/
+    # there can be multiple space separated URIs
+    # sources are specified in priority order
+    # apt does some de-duplication of sources after parsing too
+#Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
+# let's make this insecure! (jk, just testing parsing)
+Suites: noble noble-updates noble-backports
+
+Foo: Bar  # this is a separate (malformed) entry
+
+#Types: deb
+#URIs: http://security.ubuntu.com/ubuntu
+#Suites: noble-security
+#Components: main restricted universe multiverse
+#Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
+## disable security updates while we're at it
+"""
+
+ubuntu_sources_one_line = """
+deb [signed-by=/usr/share/keyrings/ubuntu-archive-keyring.gpg] http://nz.archive.ubuntu.com/ubuntu/ noble main restricted universe multiverse
+deb [signed-by=/usr/share/keyrings/ubuntu-archive-keyring.gpg] http://nz.archive.ubuntu.com/ubuntu/ noble-updates main restricted universe multiverse
+deb [signed-by=/usr/share/keyrings/ubuntu-archive-keyring.gpg] http://nz.archive.ubuntu.com/ubuntu/ noble-backports main restricted universe multiverse
+deb [signed-by=/usr/share/keyrings/ubuntu-archive-keyring.gpg] http://security.ubuntu.com/ubuntu noble-security main restricted universe multiverse
+"""
+
+
+class TestRepositoryMappingDeb822Behaviour(unittest.TestCase):
+    def test_iter_deb822_paragraphs_ubuntu_sources(self):
+        lines = ubuntu_sources_deb822.strip().split("\n")
+        paras = list(apt._iter_deb822_stanzas(lines))
```

Comment:
We're testing a lot of internal methods here. Can we take more of an approach of testing these things *via* the public interface so that the tests aren't so dependent on implementation details?